src/lib.rs

```diff
@@ -509,23 +509,31 @@ impl<T> NodeData<T> {
 
 impl<T> Drop for NodeData<T> {
     fn drop(&mut self) {
-        // Collect all descendant nodes and detach them to prevent the stack overflow.
-
-        let mut stack = Vec::new();
-        if let Some(first_child) = self.first_child.as_ref() {
-            // Create `Node` from `NodeData`.
-            let first_child = Node(first_child.clone());
-            // Iterate `self` children, without creating yet another `Node`.
-            for child1 in first_child.following_siblings() {
-                for child2 in child1.descendants() {
-                    stack.push(child2);
-                }
-            }
+        if let Some(child) = self.first_child.take() {
+            detach_all_unique(child);
         }
+    }
+}
+
+/// Detach all unique descendant and sibling nodes recursively.
+fn detach_all_unique<T>(node: Link<T>) {
+    let mut work = vec![node];
```

Comment:
`work` is a weird name. Maybe we should use something like `remaining`?

Reply:
It holds nodes to be processed, but I agree we should try to find a better name. Maybe `open` or `open_set` as in the set of items to check? The Wikipedia page for the A* search algorithm uses that. The old method used `stack` that could also work here.